### PR TITLE
Create the friendship model and its functions

### PR DESCRIPTION
**Pull request by jacobo12montes**
- Generate the friendship model(throught table)
- Create the relations on the models
- Create the controller for friendship
- Add the namespace in the config/routes.rb file
- Run and verify the migrations

### CODE REVIEW

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @jacobo12montes  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x] Remove the small empty button which is next to the name of the current user from the users list
- [x] The confirm and rehect buttons are combined in one button, fix that issue as well
- [x] Posts timeline includes posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from Pathwright)
- [x] Add unit and integration tests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from the pathwright)
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

File: app/views/posts/index.html.erb [hunk 1]
```diff
@@ -1,7 +1,12 @@
+<h1><%= current_user.name %></h1>
 <%= form_for(@post) do |form| %>
   <%= form.label :content %>
   <%= form.text_field :content, id: :post_content, class: 'form-control', placeholder: 'Add New Post' %>
-  <%= display_errors(@post) %>
+  <% if @post.errors.full_messages.any? %>
+    <p class="errors">
+      <%= "Post could not be saved. #{@post.errors.full_messages.join('. ')}" %>
+    </p>
+  <% end %>
```
Comment (acushlakoncept) [inline app/views/posts/index.html.erb]: There should be absolutely NO logic in views

File: app/views/shared/_flash_messages.html.erb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<% if notice.present? %>
+  <div class="notice">
+    <p> <%= notice %> </p>
+  </div>
+<% end %>
+<% if alert.present? %>
+  <div class="alert">
+    <p> <%= alert %> </p>
+  </div>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/shared/_flash_messages.html.erb]: There should be absolutely NO logic in views

File: app/views/shared/_friendship_status.html.erb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<% if current_user.friend?(user) %>
+  <%= link_to  "Delete", user_friendship_path(user, current_user), method: :delete, id: "user_#{user.id}",  class: "btn-cancel" %>
+<% elsif current_user.pending_friend?(user) %>
+  <%= link_to  "Accept", user_friendship_path(user, current_user), method: :put, id: "user_#{user.id}" , class: "btn-accept"  %>
+  <%= link_to  "Cancel", user_friendship_path(user, current_user), method: :delete, id: "user_#{user.id}",  class: "btn-cancel" %>
+<% elsif current_user.friend_request?(user) %>
+  <%= link_to  "Cancel", user_friendship_path(user, current_user), method: :delete, id: "user_#{user.id}",  class: "btn-cancel" %>
+<% elsif user != current_user %>
+  <%= link_to "Add", user_friendships_path(user), method: :post, class: "btn-add" %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/shared/_friendship_status.html.erb]: There should be absolutely NO logic in views

File: app/views/shared/_session_links.html.erb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<% if current_user %>
+  <%= link_to 'Sign Out', destroy_user_session_path, :method => :delete %>
+<% elsif !current_user && controller_name == 'sessions' %>
+  <%= link_to "Sign up", new_user_registration_path %><br />
+<% else %>
+  <%= link_to "Log in", user_session_path %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/shared/_session_links.html.erb]: There should be absolutely NO logic in views

File: app/views/shared/_user_posts.html.erb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<% @timeline_posts.each do |post| %>
+  <% if current_user.friend?(post.user) || current_user == post.user %>
+    <%= render post %>
+  <% end %>
+<% end %>
```
Comment (acushlakoncept) [inline app/views/shared/_user_posts.html.erb]: There should be absolutely NO logic in views

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @jacobo12montes, you have done well on this project, adding tests and implementing the previous TSE requirements.

Let's make a few changes, and we will be done.

- [x] Clicking on the see profile link produces an error. 
Screen Shot 2021-03-07 at 10 07 11 PM

- [x] Add unit test for all models, don't forget to add integration test as well

- [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase see affected files below:

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by oloomoses — CHANGES_REQUESTED** [review]
## Changes Required :recycle: 
Hello team, you have done a great job implementing the changes from the previous review. 
There are some critical issues that must be addressed before this project can be approved

### Required
- [x] Kindly Ensure that the linter and rubocop configuration is correct and all checks are passing on Github.
- [x] Some of your tests are failing, Kindly fix them.

### Optional
- [ ] the add, confirms, and reject are working perfectly. It would be better to turn them into buttons for user-friendliness.

Kindly address the above issues before submitting another review.
Feel free to leave a message on slack in case of doubts or concerns.

_slack: Oloo Moses_

**Comment by jacobo12montes** [comment]
Add the missing test and fix the failing test

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @jacobo12montes**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [x] Please make sure to setup your linters probably, check if they pass, and request another code review.


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Changes Required ⚠️**

<hr>

Hello there 👋
This is a great effort you're doing in this project so far. However, I'm afraid that the linter isn't well-configured yet. you should see the checks passing in the PR down below. Currently, I cannot see that, so kindly implement this issue to get the milestone approved 😊
good luck and happy coding 👍

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
**Hi Juan!
Great job so far:clap::clap:. The app is working as expected. However, we need to fix the failing linter errors to get this project approved:muscle:.**
**To improve**

- [ ] As highlighted by the last reviewer. Make sure the linter checks are all passing before submitting the PR for re-review. Kindly fix the linter errors.

 [Optional ]

- Kindly move the logic in application.html.erb under layout in view to a helper. There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. Check this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) on how to use a helper in rails. See [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright.

Make the above required changes and when you are ready, request another review.
Do have questions/comments? kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan

**Review by oloomoses — APPROVED** [review]
# Congratulations, Milestone Approved. :muscle: :tada: :clap: 
**Great job working on the suggested changes from the previous review.
You have passed all the requirements for this milestone :muscle: 
Feel free to merge and close this pull request.**


**I wish you the best in your next project and see you at the top :rocket:** 
#### Happy Coding
_Slack: Oloo Moses_

**Pull request merged**